proposals/p2015.md

```diff
@@ -0,0 +1,232 @@
+other options such as `IntN` (for example, `i16` versus `Int16`).
+
+### Non-goals
+
+-   This does not address any considerations around the `bool` type
+-   This does not provide a formal plan for the shape or mapping of the
+    underlying types
+    ([#767 comments](https://github.com/carbon-language/carbon-lang/issues/767#issuecomment-1214153375))
+-   This does not prescribe an official grammar for parsing of these types
+
+## Details
+
+### Syntax
+
+The syntax for a two’s complement signed integer, the unsigned integer, and the
+floating-point number corresponds to a lowercase ‘i’, ‘u’, or ‘f’ character,
+respectively, indicating the type followed by a numeric value specifying the
+width.
+
+As a regular expression, this can be illustrated as:
+
+```re
+([iuf])([1-9][0-9]*)
+```
+
+Capture group 1 indicates either an ‘i’ for a two’s complement signed integer
+type, a ‘u’ for an unsigned integer type, or a ‘f’ for a floating-point number
+type. And capture group 2 specifies the width in bits.
+
+Examples of this syntax include:
+
+-   `i32` - A 32-bit two’s complement signed integer type
+-   `u16` - A 16-bit unsigned integer type
+-   `f128` - A 128-bit IEEE 754 floating-point number type
+
+### Usage
+
+```carbon
+package sample api;
+
+fn Sum(x: i32, y: i32) -> i32 {
+  return x + y;
+}
+
+fn Main() -> i32 {
+  return Sum(4, 2);
+}
+```
+
+In the above example, `Sum` has parameters `x` and `y`, each of which is typed
+as a 32-bit two’s complement signed integer. `Main` then returns the output of
+`Sum` as a 32-bit two’s complement signed integer.
+
+### Syntax examples and considerations
+
+These examples are not meant to be formal mappings. They are only used to
+illustrate to the reader how the numeric type syntax may connect to the
+underlying language semantics in future numeric type additions.
+
+#### Syntax example use in common fixed-size numeric types
+
+[As proposed by @chandlerc](https://github.com/carbon-language/carbon-lang/issues/543#issuecomment-845620894)
+, our heuristic for type qualification here will be "Any truly pervasive
+non-aggregate/composite/compound type that is also fundamentally
+size-parameterized."
+
+These fixed-width sizes can naturally cover the common register widths of 8-,
+16-, 32-, 64-, and 128-bits, as seen in other languages.
+
+Which expand to:
+
+-   `i8`, `i16`, `i32`, `i64`, and `i128` keyword aliases for
+    `Carbon.Int(N={8,16,32,64,128})`
+-   `u8`, `u16`, `u32`, `u64`, and `u128` keyword aliases for
+    `Carbon.UInt(N={8,16,32,64,128})`
+-   `f16`, `f32`, `f64`, and `f128` keyword aliases for
+    `Carbon.Float(N={16,32,64,128})`
+
+#### Syntax example use in arbitrary fixed-size numeric types
+
+In the more general case, this syntax can express N-width types.
+
+-   `iN` for `Carbon.Int(N)`, An N-width two’s complement signed integer type
+-   `uN` for `Carbon.UInt(N)`, An N-width unsigned integer type
+-   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 floating-point number type
```

Comment:
I agree with this as well. As you stated, limiting to common sizes removes a bunch of visual confusion in the type names and only affects relatively fringe type cases. Do you think this should be reflected here in this proposal at all? Perhaps a note around this section? Or would it be more appropriate to remove it completely?

I feel like as this is a proposal centered around the syntax, it's not unreasonable to have this material covered here for the future, as long as it's clear to the reader.

Reply:
I'd suggest starting with multiples of 8 for integer types, but allowing all multiples of 8 (other than 0). LLVM has really good support for this in most operations, it only struggles with things like division and a few other oddities. We should (IMO) consider adding limits when we need to rather than front-loading them here, and I think that restriction should be on the aliased thing (`Carbon.Int(102400)` or however its spelled), not on the literal syntax.

I'd actually suggest following the same pattern for floating point *literals* as well. While we'll immediately need to restrict the set of `Carbon.Float(N)` types to the values of `N` we can implement, the restriction should be there in the library and not in the literal syntax IMO.